tutorials/helper_functions.py

```diff
@@ -0,0 +1,222 @@
+
+
+def concatenate_folder(dir_path: Path, output_filename: str):
+    """Concatenate one folder of GPS- or communication-related summaries"""
+
+    # initialize dataframe list
+    df_list = []
+
+    # loop through files in dir_path
+    for file in os.listdir(dir_path):
+        # obtain subject study_id
+        file_dir = os.path.join(dir_path, file)
+        subject_id = os.path.basename(file_dir)[:-4]
+        if file.endswith(".csv"):
+            temp_df = pd.read_csv(file_dir)
+            temp_df.insert(loc=0, column="Beiwe_ID", value=subject_id)
+            df_list.append(temp_df)
+
+    if len(df_list) > 0:
+
+        # concatenate dataframes within list --> Final Data for trajectories
+        response_data = pd.concat(df_list, axis=0)
+
+        # make directory
+        os.makedirs(dir_path / "concatenated", exist_ok=True)
+        path_resp = os.path.join(dir_path / "concatenated", output_filename)
+
+        # write to csv
+        response_data.to_csv(path_resp, index=False)
+        print(
+            "Concatenated folder " + str(dir_path)
+            + " to " + str(output_filename)
+        )
+    else:
+        print("No input data found in folder " + str(dir_path))
+
+
+def download_data(
+    keyring,
+    study_id,
+    download_folder,
+    users=None,
+    time_start="2008-01-01",
+    time_end=None,
+    data_streams=None,
+):
+    """
+    Downloads all data for specified users, time frame, and data streams.
+
+    This function downloads all data for selected users,
+    time frame, and data streams, and writes them to an
+    output folder, with one subfolder for each user, and subfolders
+    inside the user's folder for each data stream.
+    If a server failure happens, the function re-attempts the download.
+
+    Args:
+        keyring: a keyring generated by mano.keyring
+
+        users(iterable): A list of users to download data for.
+            If none are entered, it attempts to download data for all users
+
+        study_id(str): The id of a study
+
+        download_folder(str): path to a folder to download data
+
+        time_start(str): The initial date to download data
+            (Formatted in YYYY-MM-DD). Default is 2008-01-01, which is
+            before any Beiwe data existed.
+
+        time_end(str): The date to end downloads.
+            The default is today at midnight.
+
+        data_streams(iterable): A list of all data streams to download.
+            The default (None) is all possible data streams.
+
+    """
+    if study_id == "":
+        print("Error: Study ID is blank")
+        return
+
+    if (
+        keyring["USERNAME"] == ""
+        or keyring["PASSWORD"] == ""
+        or keyring["ACCESS_KEY"] == ""
+        or keyring["SECRET_KEY"] == ""
+    ):
+        print("Error: Did you set up the keyring_studies.py file?")
+        return
+
+    if not os.path.isdir(download_folder):
+        os.makedirs(download_folder, exist_ok=True)
```

Comment:
The point of `exist_ok=True` is that you don't need to check if a directory exists.
https://docs.python.org/3.8/library/os.html#os.makedirs